tests/Aql/ReturnExecutorTest.cpp

```diff
@@ -0,0 +1,182 @@
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Jan Christoph Uhde
+////////////////////////////////////////////////////////////////////////////////
+
+#include "BlockFetcherHelper.h"
+#include "catch.hpp"
+
+#include "Aql/AqlItemBlock.h"
+#include "Aql/AqlItemBlockShell.h"
+#include "Aql/ExecutionBlockImpl.h"
+#include "Aql/ExecutorInfos.h"
+#include "Aql/InputAqlItemRow.h"
+#include "Aql/ResourceUsage.h"
+#include "Aql/ReturnExecutor.h"
+#include "Aql/SingleRowFetcher.h"
+
+#include <velocypack/Builder.h>
+#include <velocypack/velocypack-aliases.h>
+
+using namespace arangodb;
+using namespace arangodb::aql;
+
+namespace arangodb {
+namespace tests {
+namespace aql {
+
+SCENARIO("ReturnExecutor", "[AQL][EXECUTOR][RETURN]") {
+  ExecutionState state;
+
+  ResourceMonitor monitor;
+  AqlItemBlockManager itemBlockManager(&monitor);
+  auto block = std::make_unique<AqlItemBlock>(&monitor, 1000, 1);
+  auto outputRegisters = make_shared_unordered_set({0});
+  auto registersToKeep = make_shared_unordered_set();
+  auto outputBlockShell =
+      std::make_unique<OutputAqlItemBlockShell>(itemBlockManager, std::move(block),
+                                                outputRegisters, registersToKeep);
+
+  REQUIRE(outputRegisters->size()  == 1);
+  REQUIRE(( *(outputRegisters->begin())  == 0 ));
+
+  RegisterId inputRegister(0);
+  RegisterId outputRegister(*(outputRegisters->begin()));
+
+  ReturnExecutorInfos infos(inputRegister, outputRegister, 1 /*nr in*/ , 1 /*nr out*/,
+                            std::unordered_set<RegisterId>{} /*to clear*/, false /*return inherit*/);
+
+  GIVEN("there are no rows upstream") {
+    VPackBuilder input;
+
+    WHEN("the producer does not wait") {
+      SingleRowFetcherHelper fetcher(input.steal(), false);
+      ReturnExecutor testee(fetcher, infos);
+      CountStats stats{};
+
+      THEN("the executor should return DONE with nullptr") {
+        OutputAqlItemRow result(std::move(outputBlockShell));
+        std::tie(state, stats) = testee.produceRow(result);
+        REQUIRE(state == ExecutionState::DONE);
+        REQUIRE(!result.produced());
+      }
+    }
+
+    WHEN("the producer waits") {
+      SingleRowFetcherHelper fetcher(input.steal(), true);
+      ReturnExecutor testee(fetcher, infos);
+      CountStats stats{};
+
+      THEN("the executor should first return WAIT with nullptr") {
+        OutputAqlItemRow result(std::move(outputBlockShell));
+        std::tie(state, stats) = testee.produceRow(result);
+        REQUIRE(state == ExecutionState::WAITING);
+        REQUIRE(!result.produced());
+
+        AND_THEN("the executor should return DONE with nullptr") {
+          std::tie(state, stats) = testee.produceRow(result);
+          REQUIRE(state == ExecutionState::DONE);
+          REQUIRE(!result.produced());
+        }
+      }
+    }
+  }
+
+  GIVEN("there are rows in the upstream") {
+    auto input = VPackParser::fromJson(
```

Comment:
Test is almost good.
Please Validate that the input is indeed returned!

Reply:
fixed